### ISSUE TITLE
Reported error with constrained TypeVars

### ISSUE THREAD

**Issue by amotzop** [microsoft/pyright#10182]
**Describe the bug**
In the following code, pyright report the following error
```
/home/asd/Desktop/test/test.py
  /home/asd/Desktop/test/test.py:20:20 - error: Argument of type "ClsA* | ClsB*" cannot be assigned to parameter "other" of type "ClsA*" in function "red"
    Type "ClsA* | ClsB*" is not assignable to type "ClsA*"
      "ClsB*" is not assignable to "ClsA*" (reportArgumentType)
  /home/asd/Desktop/test/test.py:20:20 - error: Argument of type "ClsA* | ClsB*" cannot be assigned to parameter "other" of type "ClsB*" in function "red"
    Type "ClsA* | ClsB*" is not assignable to type "ClsB*"
      "ClsA*" is not assignable to "ClsB*" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```
mypy seems to be OK with this code and doesn't raise any errors.

**Code or Screenshots**

```python
from typing import TypeVar, Self

class ClsA:
    def trans(self) -> Self:
        return self

    def red(self, other: Self) -> Self:
        return other

class ClsB:
    def trans(self) -> Self:
        return self

    def red(self, other: Self) -> Self:
        return other

T = TypeVar("T", ClsA, ClsB)

def func(arg: T) -> T:
    return arg.red(arg.trans())
```

**VS Code extension or command-line**
This happens both in VSCode and in the CLI. pyright version 1.1.398

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Value-constrained type variables are not well defined in the typing spec, and I'm not hopeful that we'll ever be able to standardize their behavior in a way that composes well with other features of the type system. You will therefore see differences in behavior between type checkers. I discuss this [here](https://microsoft.github.io/pyright/#/mypy-comparison?id=value-constrained-type-variables). 

To my knowledge, no other programming language supports value-constrained type variables, and for good reason. I recommend avoiding the use of value-constrained type variables if you can help it. There's almost always a better approach. Here's one such approach. This is the pattern that would typically be used in other languages, and it's guaranteed to work across all Python type checkers because the behaviors are well specified by the typing spec.

```python
class SupportsTransRed(Protocol):
    def trans(self) -> Self: ...

    def red(self, other: Self) -> Self: ...

def func[T: SupportsTransRed](arg: T) -> T:
    return arg.red(arg.trans())
```
